### DISCUSSION TITLE
Set required setting of a field in twig

### DISCUSSION THREAD

**Discussion by JulieVelghe** [verbb/formie#2038] [Q&A]
Hi all,
I was wondering if it is possible to override the required setting of a field (specifically an upload-field).
I have a form that is being used to create or update a portfolio entry. When there isn't an id-parameter, the form concers a new entry and some fields are required.
When there is an id-parameter, the form gets prefilled and the entry gets updated. I would like to make some of the file-upload field non-mandatory, to prevent the user from having to upload the same file again.
Any idea how I could do this? I've tried:

```
    {% do form.setFieldSettings('workFeaturedAudio', {
        'required': 'false'
        })
    %}
```
But that doesn't seem to work, as the field still ends up being required.

And an additional question, how do I then make sure that, when no new file is being provided, the file in the entry remains as is?
In the past, when we tried setting the file-field to optional. But leaving it empty would result in the file being removed from the entry. We would like to avoid that, and simply keep the existing file in the entry when no new one is being provided in the form.

**Comment by engram-design** [comment]
Using `setFieldSettings()` to change the required state should work. I believe It's an issue if you're passing a string `'false'` instead of just `false`.

```twig
{% do form.setFieldSettings('workFeaturedAudio', {
    required: true,
}) %}

{% do form.setFieldSettings('workFeaturedAudio', {
    required: false,
}) %}
```

Both the above work on my end.

As for File Upload fields, they're a little special because of how the browser handles `<input type="file">` fields. They can't be populated with a value, so Formie will check if you've actually included a value, and if not, won't touch already-uploaded files. Unlike other, similar fields like Element fields, if you were to submit the form with no values selected, but have previously selected them and stored against the submission - they'll be cleared.

But it sounds like you might also be hooking Formie up to an Entry integration - is that correct? I just want to ensure that we have our submission (the Formie side) and entry (the Craft side) terminology correct.

**Reply by JulieVelghe** [reply]
Indeed, we have an entry integration on the form where the file-upload fields are mapped to file-fields in the entry.
We also have a custom module at our disposal in case any code is better of handled there.

**Reply by JulieVelghe** [reply]
@engram-design I think that the front-end validation code now works, but I'm still getting an error that the upload file cannot be empty. I have a feeling that that might be because of the entry integration? Below is a screenshot of our mapping:

![image](https://github.com/user-attachments/assets/930cf0a3-6daf-446b-9856-a4855a4442d3)


I've also tried to change the rules on the submission via this code:

```
        Event::on(Submission::class, Submission::EVENT_DEFINE_RULES, function(SubmissionRulesEvent $event) {
            $form = $event->submission->getForm();
            $submission = $event->submission;
            if ($form->handle == "formUserAddWorkAudio") {
                $event->rules[] = [['field:workFeaturedAudio'], 'required', 'when' => function($model) {
                    return !$model->workEntryId;
                }];
            }
            if ($form->handle == "formUserAddWorkVideo") {
                $event->rules[] = [['field:workFeaturedVideo'], 'required', 'when' => function($model) {
                    return !$model->workEntryId;
                }];
            }
        });
```
I then tried to get into the payload event via this code:

```
        Event::on(Entry::class, Entry::EVENT_BEFORE_SEND_PAYLOAD, function(SendIntegrationPayloadEvent $event){
            // $payload will be entry element
            $payload = $event->payload;

            $submission = $event->submission;
            $form = $submission->getForm();

            if ($form && $form->handle == "formUserAddWorkAudio") {
                //remove the workFeaturedAudio field from the payload
                Craft::info(
                    Craft::t(
                        'ritcsfolio-module',
                        'dump the audio payload',
                        ['name' => 'Ritcsfolio Module']
                    ),
                    __METHOD__
                );
            }
            if ($form && $form->handle == "formUserAddWorkVideo") {
                //remove the workFeaturedVideo field from the payload
                Craft::info(
                    Craft::t(
                        'ritcsfolio-module',
                        'dump the video payload',
                        ['name' => 'Ritcsfolio Module']
                    ),
                    __METHOD__
                );
            }

        });
```

But when I'm checking the logs, I don't even see the message. I do see an error:

```
2024-08-27 16:10:35 [web.INFO] [formie] Submission triggered for formUserAddWorkAudio. {"memory":29717256} 
2024-08-27 16:10:35 [web.ERROR] [formie] Couldn’t save submission due to errors - {"workFeaturedAudio":["field.label.workFeaturedAudio mag niet leeg zijn."]}. {"memory":40687224} 
2024-08-27 16:10:35 [web.INFO] [application] Request context:
$_GET = [
    'p' => 'user/account/werk/37487'
]

$_POST = [
    'CRAFT_CSRF_TOKEN' => '••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••'
    'action' => 'formie/submissions/submit'
    'submitAction' => 'submit'
    'handle' => 'formUserAddWorkAudio'
    'siteId' => '4'
    'fields' => [
        'workIsEnabled' => '1'
        'fieldLabelWorkisenabled' => ''
        'workTitle' => 'Testje audio'
        'workIntroduction' => 'dit is een test tralalalala'
        'workBody' => ''
        'workCourse' => [
            0 => '6843'
        ]
        'workCategory' => [
            0 => '3081'
        ]
        'workStudyYear' => '2023'
        'workIsGraduationProject' => ''
        'workCredits' => ''
        'workEntryId' => '37487'
        'userProfileId' => '37482'
    ]
]

$_FILES = [
    'fields' => [
        'name' => [
            'workFeaturedAudio' => [
                0 => ''
            ]
            'workImages' => [
                0 => ''
            ]
            'workDocuments' => [
                0 => ''
            ]
        ]
        'full_path' => [
            'workFeaturedAudio' => [
                0 => ''
            ]
            'workImages' => [
                0 => ''
            ]
            'workDocuments' => [
                0 => ''
            ]
        ]
        'type' => [
            'workFeaturedAudio' => [
                0 => ''
            ]
            'workImages' => [
                0 => ''
            ]
            'workDocuments' => [
                0 => ''
            ]
        ]
        'tmp_name' => [
            'workFeaturedAudio' => [
                0 => ''
            ]
            'workImages' => [
                0 => ''
            ]
            'workDocuments' => [
                0 => ''
            ]
        ]
        'error' => [
            'workFeaturedAudio' => [
                0 => 4
            ]
            'workImages' => [
                0 => 4
            ]
            'workDocuments' => [
                0 => 4
            ]
        ]
        'size' => [
            'workFeaturedAudio' => [
                0 => 0
            ]
            'workImages' => [
                0 => 0
            ]
            'workDocuments' => [
                0 => 0
            ]
        ]
    ]
]

$_COOKIE = [
    '98c50eb324c847f322a5e180364464fa_username' => '05df2eaabd2b8d490033a4636018727c68366a9935a21f05089a7d52a2f37ee9a:2:{i:0;s:41:\"98c50eb324c847f322a5e180364464fa_username\";i:1;s:18:\"julie@karakters.be\";}'
    'CookieConsent' => '{stamp:\'Ryjny/n6pMezvhaP16Ka/z11VXAC0V5XnJ37jH407mXzwtPkSpLQQQ==\',necessary:true,preferences:true,statistics:true,marketing:true,method:\'explicit\',ver:1,utc:1722239190110,region:\'be\'}'
    'RitcsSessionId' => 'f565cabad14fc5c25f2d48404be3c3b1'
    '98c50eb324c847f322a5e180364464fa_identity' => '97559e61ee719e6f05755fe1295793a85a5c896b000a4778d2e6658a69863a56a:2:{i:0;s:41:\"98c50eb324c847f322a5e180364464fa_identity\";i:1;s:163:\"[37482,\"[\\\"z4VB5_7nYrTDL_aHe9hiTAJkQ9kAt7G91T1-G8GUuAHiOyMg3UHAFY-0kNFweLnGKtYwXeBfcBOFCdDbDEY4Pnq1d74K6w_HZUfE\\\",null,\\\"a565ccc5e7018c4ec7bec64e38db2966\\\"]\",3600]\";}'
    'CRAFT_CSRF_TOKEN' => '••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••'
]

$_SESSION = [
    'c49cc265eacc5f46564da2ff5374151c__flash' => [
        'formie.37455:error' => 1
    ]
    'formie:318:snapshot' => [
        'form' => [
            'collectIp' => false
            'displayCurrentPageTitle' => false
            'displayFormTitle' => false
            'displayPageProgress' => false
            'displayPageTabs' => true
        ]
    ]
    'formie:37465:snapshot' => [
        'form' => [
            'redirectUrl' => 'https://local.ritcsfolio.be/user/account'
        ]
    ]
    '98c50eb324c847f322a5e180364464fa__token' => '••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••'
    '98c50eb324c847f322a5e180364464fa__id' => 37482
    '__authKey' => '••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••'
    '98c50eb324c847f322a5e180364464fa__expire' => 1724771433
    '__duration' => 3600
    'formie:37455:snapshot' => [
        'form' => [
            'collectIp' => false
            'displayCurrentPageTitle' => false
            'displayFormTitle' => false
            'displayPageProgress' => false
            'displayPageTabs' => true
            'submitAction' => 'message'
            'submitActionFormHide' => false
            'submitActionMessage' => [
                0 => [
                    'type' => 'paragraph'
                    'content' => [
                        0 => [
                            'type' => 'text'
                            'text' => 'Jouw wijzigingen zijn opgeslagen.'
                        ]
                    ]
                ]
            ]
        ]
        'fields' => [
            'workFeaturedAudio' => [
                'required' => false
            ]
        ]
    ]
    'formie:37455:pageId' => 1731
    'formie.37455:error' => '<p>message.submission.error</p>'
]

```
Could this be because the field is required in the entry? (as is visible in the mapping screenshot)

Is there a way, perhaps via a module, to allow this upload field to be empty in the submission, and if it's empty, to ignore it from an Entry integration point of view?

**Reply by engram-design** [reply]
Sure, in the payload event, you can modify anything that's being sent to the Entry. See [beforeSendPayload](https://verbb.io/craft-plugins/formie/docs/developers/events#the-beforeSendPayload-event)

```php
use verbb\formie\events\SendIntegrationPayloadEvent;
use verbb\formie\integrations\elements\Entry;
use yii\base\Event;

Event::on(Entry::class, Entry::EVENT_BEFORE_SEND_PAYLOAD, function(SendIntegrationPayloadEvent $event) {
    $entry = $event->payload;
    $submission = $event->submission;

    // Add your logic to modify the entry's Assets field here...

    $event->payload = $entry;
});
```

**Reply by JulieVelghe** [reply]
@engram-design problem is, the code is not even getting in the payload event. I'm getting the error that is in my comment above, and the payload event doesn't seem to be triggered.

**Reply by engram-design** [reply]
Ah right, it's probably not getting to that stage yet, and it's failing on form submission. But don't forget that payload event is run in the context of a queue (by default) so it might be difficult to debug.

But, if you're sure that it's not reaching that far, I was pondering whether you could use Formie's conditionals to show/hide the field based on your logic. Even if that's populating a hidden field, which the File Upload field references in its conditions. A required hidden field will not be required, so in the form builder you could have the File Upload field set to required, but conditionally hidden.

Otherwise, I'll have to spin up an example and dive in. I'm probably not surprised there's issues programmatically changing the required state on the fly, so I'll have to take a deeper look into solutions.

**Reply by JulieVelghe** [reply]
@engram-design I don't think hidding the field is a usefull thing to do in my case. Because a user should be able to select another file, in case they want to update that portfolio-entry. But it shouldn't be mandatory, if, let's say, they only want to update a description field.
And in case it is a new entry (the code knows this because of a hidden field that gets populated), then the file-field should be mandatory.

**Reply by JulieVelghe** [reply]
@engram-design using your proposition of hidding the field might be a route. I've added a checkbox to ask the user if he/she want to upload a different file. If unchecked, the file upload field is not shown in the form, and I can save the form, the entry gets updated (I tried with a description field), but the linked file in the entry remained the same. Exactly what I wanted.
I'll have to do some more tests, but this might be a workable solution.

**Reply by engram-design** [reply]
Yeah that might be a nice alternative. This also goes without saying that one of the core issues here is the native `<input type="file">` element being extremely limited in these sorts of things, which will be addressed in https://github.com/verbb/formie/issues/133
